src-control/Control/RefCount.hs

```diff
@@ -476,42 +483,51 @@ finaliserRefTracker inner refid allocSite = do
           Just (refid', _) | refid < refid' -> return ()
           _ -> writeIORef globalForgottenRef (Just (refid, allocSite))
 
-assertNoForgottenRefs :: IO ()
-assertNoForgottenRefs = do
+getForgottenRef :: IO (Maybe RefException)
+getForgottenRef = do
     mrefs <- readIORef globalForgottenRef
     case mrefs of
-      Nothing                -> return ()
+      Nothing                -> return Nothing
       Just (refid, allocSite) -> do
-        -- Clear the var so we don't assert again.
+        -- Clear the var so we don't get the same reference again.
         writeIORef globalForgottenRef Nothing
-        throwIO (RefNeverReleased refid allocSite)
+        pure (Just (RefNeverReleased refid allocSite))
+
+throwIfJust :: (MonadThrow m, Exception e) => Maybe e -> m ()
+throwIfJust = mapM_ throwIO
+
+_assertNoForgottenRefs :: (PrimMonad m, MonadThrow m) => m ()
+_assertNoForgottenRefs =
+    throwIfJust =<< do
+    unsafeIOToPrim getForgottenRef
 
-assertNoUseAfterRelease :: (PrimMonad m, HasCallStack) => Ref a -> m ()
+assertNoUseAfterRelease :: (PrimMonad m, MonadThrow m, HasCallStack) => Ref a -> m ()
 assertNoUseAfterRelease Ref { reftracker = RefTracker refid _weak outer allocSite } =
+  throwIfJust =<< do
```

Comment:
This complexity comes from the approach of moving the exception throwing outside of the `unsafeIOToPrim`, which means we have to return and then throw instead.

But instead we could just move the `unsafeIOToPrim` inwards to the couple places it's actually needed: the `readIORef` operations:
```
    released <- unsafeIOToPrim (readIORef =<< readIORef outer)
```
and then we can just use normal `throwIO` from `MonadThrow`

Reply:
Done